Automattic/wp-calypso
# Gutenboarding: Switch plans grid to use plans/detail endpoint to fetch plans details via API

This is a variant of #46478 to explore using a (new) plans/detail endpoint to power the plans details and comparison feature information in the plans grid, used in Gutenboarding and in the launch flow in the Editing Toolkit.

#### Todos before this is ready for review

* [x] Group together the different storage features (they currently each get their own line), and support non-checkbox types of features in the list (this is currently the only one)
* [x] Why is the checkmark not displaying for Jetpack essential features? (Fixed, needed to add initial value to Array.reduce call)
* [x] Clean up names and types
* [x] Add loading state, e.g. pulsing rows like in the domains picker
* [x] Pre-load the plans details before the user reaches the step (implemented in https://github.com/Automattic/wp-calypso/pull/46588/commits/3b09259c2053a946fe8838222df80321382ff42b)
* [x] ~Selecting a plan via URL doesn't currently work because the plans haven't loaded yet. E.g. go to wordpress.com/new/design/premium and it should select the premium plan, however it currently doesn't work in this branch.~ Fixed in https://github.com/Automattic/wp-calypso/pull/46588/commits/3b09259c2053a946fe8838222df80321382ff42b by ensuring we load the plans early
* [ ] Add error handling? (We don't _yet_ have error handling when getting prices, so do we need it here? Ideally maybe a notice if a network request fails, and a button to re-request?)
* [x] This has not yet been tested in the launch flow in the Editing Toolkit / after a site has been created

#### Changes proposed in this Pull Request

* Update `data-stores/src/plans` to support pulling plans details from the `plans/details` API endpoint.
* Update `data-stores/src/plans` reducers to include plans, features, and featuresByType.
* Remove `plans-data` from `data-stores/src/plans`, and retain required constants in `constants.ts`.
* Update the `plans-grid` package to support a loading UI state when the plans / plans details are empty.
* Adjust the logic in `plans-grid/src/plans-details` to iterate over a list of features by id, and look up the feature details by object key.
* Note that #46557 removed importing `@automattic/react-i18n` to swap it out for `@wordpress/i18n` — this PR currently adds back in `@automattic/react-i18n` so that we can get the current locale via `i18nLocale` from `useI18n()`.

#### Screenshot

Plans UI loading state from a reload on the plans modal page:

![plans-loading-state-small](https://user-images.githubusercontent.com/14988353/97399861-22f5c480-1942-11eb-9196-8c0c529c661a.gif)

#### Testing instructions



* ~Apply D51380-code~ (plans/details endpoint is now merged)

* Go to calypso.localhost:3000/new and go to the plans step or the plans modal, e.g. http://calypso.localhost:3000/new/plans-modal/ or a translated variant, e.g. http://calypso.localhost:3000/new/plans-modal/ko
* Most of the strings are not translated yet, however you should see that _some_ of them are
* Check that the plans and their features match what's on HEAD
* Check that the feature grid / detailed comparison matches what's on HEAD
* Reload the page and check that the loading UI state matches up with the final UI state once the plans details have loaded in
* Test the loading UI state across browsers / mobile devices to make sure the styling looks okay

##### Make sure that plans selected by URL work as expected:

* From an incognito session WITH COOKIES ALLOWED, go to calypso.localhost:3000/new/ecommerce and after entering your site title, make sure that the ecommerce plan is selected (it should say so in the top right hand corner of the UI). Complete the signup flow and you shouldn't be prompted to select a plan, but you _should_ be prompted to create an account
* Create an account and make sure you're still redirected to the checkout as expected

##### Make sure it works correctly in the Editing Toolkit / launch flow

* Work out how to test this in the launch flow in the Editing Toolkit, since there are no changes to the Editing Toolkit directly in this PR (ensure that when the Editing Toolkit is next built (?) that the plans grid functions as expected)

##### Question about the PR:

* Currently `getPlansDetails` is responsible for invoking the `getPlansDetails` resolver, which loads much of the plans state. Is this side-effect reasonable, or do we need to be more deliberate to ensure that this gets called when we expect? (For example, I had to add an explicit call to this selector to ensure that plans details loaded early in https://github.com/Automattic/wp-calypso/pull/46588/commits/3b09259c2053a946fe8838222df80321382ff42b)

Related to: #37665
